File: python_version/binary_search_tree.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None
    def getValue(self):
        return self.data
    def hasLeft(self):
        return self.left is not None
    def hasRight(self):
        return self.right is not None

class BinarySearchTree:
    def __init__(self):
        self.root = None
    def search(self, value):
        return self._search(value, self.root)
    def _search(self, value, leaf):
        if leaf is None:
            return False
        elif leaf.getValue() == value:
            return leaf
        elif value < leaf.getValue() and leaf.hasLeft():
            return self._search(value, leaf.left)
        elif value > leaf.getValue() and leaf.hasRight():
            return self._search(value, leaf.right)
        else:
            return False
    def insert(self, value):
        if not self.root:
            self.root = Node(value)
        else:
            self._insert(value, self.root)
    def _insert(self, value, leaf):
        if value < leaf.getValue():
            if leaf.hasLeft():
                self._insert(value, leaf.left)
            else:
                leaf.left = Node(value)
        else:
            if leaf.hasRight():
                self._insert(value, leaf.right)
            else:
                leaf.right = Node(value)
    def delete(self, value):
        self.root = self._delete(value, self.root)
    def _delete(self, value, leaf=None):
        if leaf is not None:
            if leaf.data == value:
                if not leaf.hasLeft() and not leaf.hasRight():
                    leaf = None
                elif leaf.hasLeft() and not leaf.hasRight():
                    leaf = leaf.left
                elif not leaf.hasLeft() and leaf.hasRight():
                    leaf = leaf.right
                else:
                    successor = leaf.left
                    parent = None
                    while successor and successor.hasRight:
                        parent = successor
                        successor = successor.right
                    leaf.data = successor.data
                    if parent:
                        parent.right = None
                    else:
                        leaf.left = None
            elif value < leaf.data and leaf.hasLeft():
                leaf.left = self._delete(value, leaf.left)
            elif value > leaf.data and leaf.hasRight():
                leaf.right = self._delete(value, leaf.right)
            return leaf
    def treesize(self):
        return self._treesize(self.root)
    def _treesize(self, leaf=None):
        if leaf is None:
            return 0
        else:
            return self._treesize(leaf.left) + 1 + self._treesize(leaf.right)
    def treedepth(self):
        return self._treedepth(self.root)
    def _treedepth(self, leaf=None):
        if leaf is None:
            return 0
        else:
            return max(self._treedepth(leaf.left), self._treedepth(leaf.right)) + 1
    def getmax(self, leaf=None):
        if leaf is None:
            return None
        else:
            if leaf.hasRight():
                return self.getmax(leaf.right)
            else:
                return leaf.getValue()
    def getmin(self, leaf=None):
        if leaf is None:
            return None
        else:
            if leaf.hasLeft():
                return self.getmin(leaf.left)
            else:
                return leaf.getValue()
    #Print level by level from top to the bottom
    def BreadthFirstTraversal(self, leaf):
        if leaf is not None:
            queue = [leaf]
            output = []
            while len(queue) > 0:
                curr = queue.pop(0)
                output.append(curr.getValue())
                if curr.hasLeft():
                    queue.append(curr.left)
                if curr.hasRight():
                    queue.append(curr.right)
            return output
    def DepthFirstTraversal(self, leaf):
        if leaf is not None:
            stack = [leaf]
            output = []
            while len(stack) > 0:
                curr = stack.pop()
                output.append(curr.getValue())
                if curr.hasLeft():
                    stack.append(curr.left)
                if curr.hasRight():
                    stack.append(curr.right)
            return output
    def printlevel(self, leaf, level):
        if level == 0:
            print leaf.data
        else:
            if leaf.hasLeft():
                self.printlevel(leaf.left, level-1)
            if leaf.hasRight():
                self.printlevel(leaf.right, level-1)
    def printleveldown(self, leaf):
        for i in xrange(self.treedepth()):
            self.printlevel(leaf, i)
    def printlevelup(self, leaf):
        for i in xrange(self.treedepth(), -1, -1):
            self.printlevel(leaf, i)
    def printdeepestlevel(self, leaf):
        queueone = [leaf]
        queuetwo = []
        printlist = []
        while len(queueone) != 0:
            node = queueone.pop(0)
            if node.left:
                queuetwo.append(node.left)
            if node.right:
                queuetwo.append(node.right)
            if len(queueone) == 0:
                if len(queuetwo) != 0:
                    printlist = []
                    for item in queuetwo:
                        queueone.append(item)
                        printlist.append(item.getValue())
                    queuetwo = []
                else:
                    print printlist

#Test cases
'''
tree = BinarySearchTree()
tree.insert(8)
tree.insert(5)
tree.insert(12)
tree.insert(4)
tree.insert(6)
tree.insert(15)
tree.insert(11)
tree.insert(29)
tree.insert(1)
print 'Before delete 6: ', tree.search(6), tree.search(6).getValue()
tree.delete(6)
print 'After delete 6: ', tree.search(6)
print 'Get Max: ', tree.getmax(tree.root)
print 'Get Min: ', tree.getmin(tree.root)
print 'Get Tree Size: ', tree.treesize()
print 'Get Tree Depth: ', tree.treedepth()
print 'BFS: ', tree.BreadthFirstTraversal(tree.root)
print 'DFS: ', tree.DepthFirstTraversal(tree.root)
print 'Print level 2'
tree.printlevel(tree.root, 2)
print 'Print Level Up'
tree.printlevelup(tree.root)
print 'Print Level Down'
tree.printleveldown(tree.root)
print 'Print the deepest Level'
tree.printdeepestlevel(tree.root)
'''

File: java_version/BinarySearchTree/BinarySearchTree.java
import java.io.*;
import java.util.*;

class BinarySearchTree{

    private Node root;

    public BinarySearchTree(){
        this.root = null;
    }

    public BinarySearchTree(int value){
        this.root = new Node(value);
    }
    
    public boolean search(int value){
        if(this.root == null){
            return false;
        }else{
            return _search(value, this.root);
        }
    }

    public boolean _search(int value, Node leaf){
        if(leaf == null){
            return false;
        }else if(leaf.getValue() == value){
            return true;
        }else if(value < leaf.getValue()){
            return _search(value, leaf.getLeft());
        }else if(value > leaf.getValue()){
            return _search(value, leaf.getRight());
        }else{
            return false;
        }
    }

    public void insert(int value){
        if(this.root == null){
            this.root = new Node(value);
        }else{
            _insert(value, this.root);
        }
    }

    private void _insert(int value, Node leaf){
        if(value < leaf.getValue()){
            if(leaf.getLeft() == null){
                leaf.setLeft(new Node(value));
            }else{
                _insert(value, leaf.getLeft());
            }
        }else{
            if(leaf.getRight() == null){
                leaf.setRight(new Node(value));
            }else{
                _insert(value, leaf.getRight());
            }
        }
    }

    public void delete(int value){
        this.root = _delete(value, this.root);
    }

    private Node _delete(int value, Node leaf){
        if(leaf == null){
            return null;
        }else{
            if(leaf.getValue() == value){
                if(leaf.getLeft() == null && leaf.getRight() == null){
                    leaf = null;
                }else if(leaf.getLeft() != null && leaf.getRight() == null){
                    leaf = leaf.getLeft();
                }else if(leaf.getLeft() == null && leaf.getRight() != null){
                    leaf = leaf.getRight();
                }else{
                    Node successor = leaf.getLeft();
                    Node parent = null;
                    while(successor != null && successor.getRight() != null){
                        parent = successor;
                        successor = successor.getRight();
                    }
                    leaf.setValue(successor.getValue());
                    if(parent != null){
                        parent.setRight(null);
                    }else{
                        leaf.setLeft(null);
                    }
                }
            }else if(value < leaf.getValue()){
                leaf.setLeft(_delete(value, leaf.getLeft()));
            }else{
                leaf.setRight(_delete(value, leaf.getRight()));
            }
            return leaf;
        }
    }

    public int treeSize(){
        return _treeSize(this.root);
    }

    private int _treeSize(Node leaf){
        if(leaf == null){
            return 0;
        }else{
            return _treeSize(leaf.getLeft()) + 1 + _treeSize(leaf.getRight());
        }
    }

    public int treeDepth(){
        return _treeDepth(this.root);
    }

    private int _treeDepth(Node leaf){
        if(leaf == null){
            return 0;
        }else{
            return Math.max(_treeDepth(leaf.getLeft()), _treeDepth(leaf.getRight())) + 1;
        }
    }

    public int getMax(){
        if(this.root == null){
            return 0;
        }
        Node temp = this.root;
        while(temp.getRight() != null){
            temp = temp.getRight();
        }
        return temp.getValue();
    }

    public int getMin(){
        return _getMin(this.root);
    }

    public int _getMin(Node leaf){
        if(leaf == null){
            return 0;
        }else{
            if(leaf.getLeft() != null){
                return _getMin(leaf.getLeft());
            }else{
                return leaf.getValue();
            }
        }
    }

    public void BreadFirstTraversal(){
        if(this.root != null){
            List<Node> queue = new ArrayList<Node> ();
            queue.add(this.root);
            List<Node> output = new ArrayList<Node> ();
            while(queue.size() > 0){
                Node curr = queue.remove(0);
                output.add(curr);
                if(curr.getLeft() != null)
                    queue.add(curr.getLeft());
                if(curr.getRight() != null)
                    queue.add(curr.getRight());
            }
            System.out.print("BFS Traversal Result: ");
            for(int i=0; i<output.size(); i++){
                System.out.print(String.valueOf(output.get(i).getValue()) + " ");
            }
            System.out.println();
        }
    }

    public void DepthFirstTraversal(){
        if(this.root != null){
            List<Node> stack = new ArrayList<Node> ();
            stack.add(this.root);
            List<Node> output = new ArrayList<Node> ();
            while(stack.size() > 0){
                Node curr = stack.remove(stack.size()-1);
                output.add(curr);
                if(curr.getLeft() != null)
                    stack.add(curr.getLeft());
                if(curr.getRight() != null)
                    stack.add(curr.getRight());
            }
            System.out.print("DFS Traversal Result: ");
            for(int i=0; i<output.size(); i++){
                System.out.print(String.valueOf(output.get(i).getValue()) + " ");
            }
            System.out.println();
        }
    }

    public void printlevel(Node leaf, int level){
        if(level == 0){
            System.out.println(leaf.getValue());
        }else{
            if(leaf.getLeft() != null){
                printlevel(leaf.getLeft(), level-1);
            }
            if(leaf.getRight() != null){
                printlevel(leaf.getRight(), level-1);
            }
        }
    }

    public void printDeepestLevel(Node leaf){
        List<Node> queueone = new ArrayList<Node> ();
        List<Node> queuetwo = new ArrayList<Node> ();
        List<Integer> printlist = new ArrayList<Integer> ();
        queueone.add(leaf);
        while(queueone.size() > 0){
            Node curr = queueone.remove(0);
            if(curr.getLeft() != null){
                queuetwo.add(curr.getLeft());
            }
            if(curr.getRight() != null){
                queuetwo.add(curr.getRight());
            }
            if(queueone.size() == 0){
                if(queuetwo.size() != 0){
                    printlist = new ArrayList<Integer> ();
                    for(Node node : queuetwo){
                        queueone.add(node);
                        printlist.add(node.getValue());
                    }
                    queuetwo = new ArrayList<Node> ();
                }else{
                    System.out.print("Deepest level: ");
                    for(Integer i : printlist){
                        System.out.print(String.valueOf(i) + " ");
                    }
                    System.out.println();
                }
            }
        }
    }

    public static void main(String[] args){
        List<String> strList = new ArrayList<String>(Arrays.asList(args[0].split(",")));
        List<Integer> intList = new ArrayList<Integer>();
        for(String str : strList){
            intList.add(Integer.parseInt(str));
        }
        BinarySearchTree tree = new BinarySearchTree();
        for(Integer i : intList){
            tree.insert(i);
        }
        System.out.println("Tree size: " + String.valueOf(tree.treeSize()));
        System.out.println("Tree depth: " + String.valueOf(tree.treeDepth()));
        tree.printDeepestLevel(tree.root);
        tree.BreadFirstTraversal();
        tree.DepthFirstTraversal();
        if(args.length == 2){
            System.out.println("Searching: " + args[1]);
            Integer target = Integer.parseInt(args[1]);
            System.out.println(tree.search(target));
        }
        if(args.length == 3){
            System.out.println("Deleting: " + args[1]);
            tree.delete(Integer.parseInt(args[1]));
            System.out.println("Tree size: " + String.valueOf(tree.treeSize()));
            System.out.println("Searching: " + args[2]);
            System.out.println(tree.search(Integer.parseInt(args[2])));
        }
    }
}
